// openfsm-user-dto/UserDTO.js
class UserDTO {
  constructor(user = null ) {
    if(!user) return null;
    this._id = Number(user.id ?? null) ;
    this._email = user.email;
    this._password = user.password;
    this._name = user.name;
    this._created = user.created;
    this._updated = user.updated;
    this._blocked = user.blocked;
    this._confirmed = user.confirmed;    
    this._verificationCode = user.verification_code;
  }

  // Getters
  getId() {
    return this._id;
  }

  getEmail() {
    return this._email;
  }

  getPassword() {
    return this._password;
  }

  getName() {
    return this._name;
  }

  getCreated() {
    return this._created;
  }

  getUpdated() {
    return this._updated;
  }

  getBlocked() {
    return this._blocked;
  }

  getConfirmed() {
    return this._confirmed;
  }

  getVerificationCode() {
    return this._verificationCode;
  }
  

  // Setters
  setId(id) {
    this._id = id;
  }

  setEmail(email) {
    this._email = email;
  }

  setPassword(password) {
    this._password = password;
  }

  setName(name) {
    this._name = name;
  }

  setCreated(created) {
    this._created = created;
  }

  setUpdated(updated) {
    this._updated = updated;
  }

  setBlocked(blocked) {
    this._blocked = blocked;
  }

  setConfirmed(confirmed) {
    this._confirmed = confirmed;
  }

  setVerificationCode(verificationCode) {
    this._verificationCode = verificationCode;
  }
}

module.exports = { UserDTO };
